### ISSUE TITLE
Some problems were encountered during the training process

### ISSUE THREAD

**Issue by zhuyf66** [mzf666/MATPO#6]
This is meaningful work!

When I'm getting ready for training, I need to prepare the corresponding search engine. Does the author use Serper exclusively for the search? Given the high cost of search, are there any more cost-effective search tools to use?

When I started training, I used the script provided by the author to train on the qwen3_8b model, but the training curve kept fluctuating and failed to improve convergence. Has the author ever encountered this situation? What good suggestions are there for solving this problem?

**Comment by mzf666** [comment]
Dear Zhuyf66, thanks for your interest.

Q1: Do we use Serper exclusively?
A1: The current repo does use Serper exclusively. You can switch to other searching tools by wrapping your desired search service into a new MCP tool, following the instructions [here](https://github.com/mzf666/MATPO?tab=readme-ov-file#adding-new-mcp-tools-to-the-planner-agent).

Q2: How to improve training stability?
A2: In addition to the observations shared in our [paper](https://arxiv.org/abs/2510.04678), we have following suggestions:
- **Use larger models**, e.g. Qwen3-14B: small models may fail to handle and understand long search-and-think trajectories, which may affect the overall performance.
- **Use larger rollout size and batch sizes**: small rollout size or batch sizes may affect the policy gradient estimation for GRPO.
- **Use stronger LLM API in the LLM-as-Judge module**: stronger LLM judger helps assigning correct reward.
- **Try to adjust the training hyperparameters**, e.g. learning rate or learning scheduler for your model. We have not conducted hyper-parameter searching yet, and there exists better training configurations.
